### ISSUE TITLE
Proposal: separate Predictor objects

### ISSUE THREAD

**Issue by st--** [GPflow/GPflow#1599]
# Feature request

`model.predict_f()` currently recomputes everything (kernel matrix, Cholesky, triangular_solve) from scratch at each call. Instead of recomputing these quantities, we should provide a predictor object that only does those computations that touch the new input points.

## Motivation

Recomputing lots of cacheable quantities from scratch is inefficient and unnecessarily slow. The slowness of predictions (particularly in a loop) is a well-known, recurring issue, e.g. https://github.com/GPflow/GPflow/issues/1554, https://github.com/GPflow/GPflow/issues/1189, https://github.com/GPflow/GPflow/issues/1030, https://github.com/GPflow/GPflow/issues/1003. We recently merged https://github.com/GPflow/GPflow/pull/1528 which was added so that a special-cased version of this proposal could be implemented down-stream. (Note that there are further efficiency gains possible that even #1528 does not allow for.) It would be much better to implement this properly, and make it available for all users out of the box.

## Proposal

**Describe the solution you would like**
Ideally, `model.predictor()` (or similar) returns an (optionally `tf.function`-decorated) closure that takes new input points `Xnew` and computes the prediction efficiently (having precomputing and cached all that can be computed without access to the test inputs).

Questions:
* Should this be extended to the other predict functions (predict mean and var, predict log density)?
* How will we be able to cache all the quantities needed within the various `conditional` branches?

**Comment by pio-neil** [comment]
> Ideally, model.predictor() (or similar) returns an (optionally tf.function-decorated) closure that takes new input points Xnew and computes the prediction efficiently (having precomputing and cached all that can be computed without access to the test inputs).

Would it make sense for `model.predictor()` to return a `Predictor` class, which could then implement multiple methods (`predict_f`, `predict_y` etc).

**Comment by st--** [comment]
> Would it make sense for `model.predictor()` to return a `Predictor` class, which could then implement multiple methods (`predict_f`, `predict_y` etc)?

Yes, that would certainly be an option as well! This could be a more extensive effort, related to thoughts/discussions we've had before around a `Posterior` object (that fully decouples from the model and contains all the prediction methods normally associated with the `Model` instance).

**Comment by joelberkeley-secondmind** [comment]
this ticket is introduced as: here's a problem, implement this solution. How about: here's a problem, solve the problem, here's one possible way of doing it. So as to avoid biasing to particular solutions?

**Comment by st--** [comment]
@joelberkeley-secondmind 

> this ticket is introduced as: here's a problem, implement this solution. Can I suggest: here's a problem, solve the problem, here's one possible way of doing it. So as to avoid biasing to particular solutions?

Hah! I got biased by the issue template. Do you want to go ahead and edit the issue accordingly to make it unbiased ?

**Comment by awav** [comment]
Hello, ladies and gentlemen!

There is one feature of TF, that people not aware of and that is a _constant folding_ feature. Tensorflow has multiple levels of optimisation and constant folding is presented in two of them: Grappler and XLA. Usually, we don't apply XLA, but  Tensorflow always runs Grappler (almost). Constant folding works in the following way: Grappler walks over static operations with leaves at constants and applies heuristics to fold them into a new constant. Extending this solution allowing a broader spectrum of updates in the graph could make predict methods faster in our code without any change in Tensorflow.

Unfortunately, TF doesn't have much information about [constant folding](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/constant_folding.h), so technically the only way to learn more about it is the code reading.

**Comment by joelberkeley-secondmind** [comment]
i'd be interested to know how you handle equaiity of inputs. If I call `model.predict_f(X)` and `model.predict_f(XX)`, how close do X and XX need to be to use the cache? Exact equality would be limited by imprecision in tf operations (i.e. two points _should_ be the same but aren't exactly)
